### PR TITLE
Allow Smalldatetime type and date type compare to choose index clause.

### PR DESCRIPTION
**Pull request by Yvinayak07**
### 1. Issue:

Index scan would not be chosen by optimiser if there is any filter that involves operator between Smalldatetime and date even though index is created on Smalldatetime column. For example,

```
create table babel_4517 (
    date_col date NULL,
    datetime_col datetime NULL,
    datetime2_col datetime2 NULL,
    smalldatetime_col smalldatetime NULL
)
GO

create NONCLUSTERED INDEX smalldatetime_col_indx on babel_4517
(
    smalldatetime_col ASC
)
GO

```
Here, Index scan would not be chosen for following query as operator class between smalldatetime and date does not exist.

```
select * from babel_4517 where smalldatetime_col <= cast('2023-08-31' as date) and smalldatetime_col >= cast('2023-08-31' as date);

```

### 2. Changes made to fix the issues

We have defined an operator class between smalldatetime and date.

### Issues Resolved
Task: [BABEL-4709](https://jira.rds.a2z.com/browse/BABEL-4709)
Authored-by: yashneet vinayak [yashneet@amazon.com](mailto:yashneet@amazon.com)
Signed-off-by: yashneet vinayak [yashneet@amazon.com](mailto:yashneet@amazon.com)

### Test Scenarios Covered ###
* **Use case based -**Yes

* **Boundary conditions -** Yes

* **Arbitrary inputs -**Yes

* **Negative test cases -**

* **Minor version upgrade tests -** 

* **Major version upgrade tests -** 

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**Yes

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_common/sql/smalldatetime.sql [hunk 1]
```diff
@@ -470,6 +470,11 @@ RETURNS INT4
 AS 'timestamp_cmp'
 LANGUAGE internal IMMUTABLE STRICT PARALLEL SAFE;
 
+CREATE OR REPLACE FUNCTION  smalldatetime_date_cmp(sys.SMALLDATETIME, date)
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_common/sql/smalldatetime.sql]: nit: better to add sys. smalldatetime_date_cmp. Modify the same for smalldatetime_cmp.

File: contrib/babelfishpg_common/sql/smalldatetime.sql [hunk 2]
```diff
@@ -489,6 +494,14 @@ DEFAULT FOR TYPE sys.SMALLDATETIME USING hash AS
     OPERATOR    1   =  (sys.SMALLDATETIME, sys.SMALLDATETIME),
     FUNCTION    1   smalldatetime_hash(sys.SMALLDATETIME);
 
+ALTER OPERATOR FAMILY smalldatetime_ops USING btree ADD
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_common/sql/smalldatetime.sql]: nit: Add sys. smalldatetime_ops

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+SELECT set_config('search_path', 'sys, '||current_setting('search_path'), false);
+
+CREATE OR REPLACE FUNCTION  smalldatetime_date_cmp(sys.SMALLDATETIME, date)
+RETURNS INT4
+AS 'timestamp_cmp_date'
+LANGUAGE internal IMMUTABLE STRICT PARALLEL SAFE;
+
+ALTER OPERATOR FAMILY smalldatetime_ops USING btree ADD
+    OPERATOR    1   <  (sys.SMALLDATETIME, date),
+    OPERATOR    2   <= (sys.SMALLDATETIME, date),
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql]: Also, we can add sys.<, sys.= etc.

**Review by ahmed-shameem — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_common/sql/smalldatetime.sql (hunk 1 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/smalldatetime.sql]: Resolved.

File: contrib/babelfishpg_common/sql/smalldatetime.sql (hunk 2 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/smalldatetime.sql]: resolved.

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql (hunk 3 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql]: resolved.

File: contrib/babelfishpg_common/sql/smalldatetime.sql [hunk 4]
```diff
@@ -489,6 +494,14 @@ DEFAULT FOR TYPE sys.SMALLDATETIME USING hash AS
     OPERATOR    1   =  (sys.SMALLDATETIME, sys.SMALLDATETIME),
     FUNCTION    1   smalldatetime_hash(sys.SMALLDATETIME);
 
+ALTER OPERATOR FAMILY sys.smalldatetime_ops USING btree ADD
```
Comment (Deepesh125) [inline contrib/babelfishpg_common/sql/smalldatetime.sql]: why cant we add it as part of original create operator family command?

Comment (Deepesh125) [inline contrib/babelfishpg_common/sql/smalldatetime.sql]: This may not work if you have index on `date` column and predicate involves smalldatetime column. Please check if we have such test cases? and add if there is not already

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql [hunk 5]
```diff
@@ -0,0 +1,24 @@
+------------------------------------------------------------------------------
+---- Include changes related to other datatypes except spatial types here ----
+------------------------------------------------------------------------------
+
+-- complain if script is sourced in psql, rather than via ALTER EXTENSION
+\echo Use "ALTER EXTENSION ""babelfishpg_common"" UPDATE TO "4.5.0"" to load this file. \quit
+
+SELECT set_config('search_path', 'sys, '||current_setting('search_path'), false);
+
+CREATE OR REPLACE FUNCTION  sys.smalldatetime_date_cmp(sys.SMALLDATETIME, date)
```
Comment (Deepesh125) [inline contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql:10]: run this script repeatedly to ensure that there is not any issue with the upgrade path

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_common/sql/smalldatetime.sql [hunk 6]
```diff
@@ -482,7 +487,13 @@ DEFAULT FOR TYPE sys.SMALLDATETIME USING btree AS
...
     OPERATOR    4   >= (sys.SMALLDATETIME, sys.SMALLDATETIME),
     OPERATOR    5   >  (sys.SMALLDATETIME, sys.SMALLDATETIME),
-    FUNCTION    1   smalldatetime_cmp(sys.SMALLDATETIME, sys.SMALLDATETIME);
+    FUNCTION    1   sys.smalldatetime_cmp(sys.SMALLDATETIME, sys.SMALLDATETIME),
+    OPERATOR    1   <  (sys.SMALLDATETIME, date),
+    OPERATOR    2   <= (sys.SMALLDATETIME, date),
+    OPERATOR    3   =  (sys.SMALLDATETIME, date),
+    OPERATOR    4   >= (sys.SMALLDATETIME, date),
+    OPERATOR    5  >  (sys.SMALLDATETIME, date),
+    FUNCTION    1   sys.smalldatetime_date_cmp(sys.SMALLDATETIME, date);
```
Comment (sumitj824) [inline contrib/babelfishpg_common/sql/smalldatetime.sql]: Instead of modifying the existing operator class, I would suggest to create new operator class named like `sys.smalldatetime_date_ops`.

File: contrib/babelfishpg_common/sql/smalldatetime.sql [hunk 7]
```diff
@@ -470,6 +470,11 @@ RETURNS INT4
 AS 'timestamp_cmp'
 LANGUAGE internal IMMUTABLE STRICT PARALLEL SAFE;
 
+CREATE OR REPLACE FUNCTION  sys.smalldatetime_date_cmp(sys.SMALLDATETIME, date)
```
Comment (sumitj824) [inline contrib/babelfishpg_common/sql/smalldatetime.sql]: NIT: no need of `REPLACE` here.

File: contrib/babelfishpg_common/sql/smalldatetime.sql (hunk 4 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/smalldatetime.sql]: Yes, if we create index on date column and compare with smalldatetime will not choose index clause as their is not operator family for date. Added those test cases.

Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/smalldatetime.sql]: I have create a new operator class that adds operator to smalldatetime_ops (already exsisting family).

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql (hunk 5 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql:10]: This failed as exception handling was missing but added a exception handling block and tested locally and it works fine.

File: contrib/babelfishpg_common/sql/smalldatetime.sql (hunk 6 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/smalldatetime.sql]: Created a new operator class that adds operator to smalldatetime_ops family.

File: contrib/babelfishpg_common/sql/smalldatetime.sql (hunk 7 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/smalldatetime.sql]: Resolved

File: contrib/babelfishpg_common/sql/smalldatetime.sql [hunk 8]
```diff
@@ -482,7 +487,16 @@ DEFAULT FOR TYPE sys.SMALLDATETIME USING btree AS
...
+    FUNCTION    1   sys.smalldatetime_cmp(sys.SMALLDATETIME, sys.SMALLDATETIME);
+
+CREATE OPERATOR CLASS sys.smalldatetime_date_ops
+FOR TYPE sys.SMALLDATETIME USING btree FAMILY smalldatetime_ops AS
+    OPERATOR    1   sys.<  (sys.SMALLDATETIME, date),
+    OPERATOR    2   sys.<= (sys.SMALLDATETIME, date),
+    OPERATOR    3   sys.=  (sys.SMALLDATETIME, date),
+    OPERATOR    4   sys.>= (sys.SMALLDATETIME, date),
+    OPERATOR    5   sys.>  (sys.SMALLDATETIME, date),
+    FUNCTION    1   sys.smalldatetime_date_cmp(sys.SMALLDATETIME, date);
```
Comment (sumitj824) [inline contrib/babelfishpg_common/sql/smalldatetime.sql:504]: Are you sure that only this operator class is sufficient while dealing with between `sys.SMALLDATETIME` and `date` type? What about when left arg is `date` and right arg is `sys.SMALLDATETIME`? We would need one more operator class and compartor function for that.

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+RETURNS INT4
+AS 'timestamp_cmp_date'
+LANGUAGE internal IMMUTABLE STRICT PARALLEL SAFE;
+
+CREATE FUNCTION  sys.date_smalldatetime_cmp(date, sys.SMALLDATETIME)
+RETURNS INT4
+AS 'date_cmp_timestamp'
+LANGUAGE internal IMMUTABLE STRICT PARALLEL SAFE;
+
+-- Opartor class for smalldatetime_ops to incorporate various operator between smalldatetime and date for Index scan
```
Comment (rohit01010) [inline contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql]: NIT: ~Opartor~ Operator

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    OPERATOR    1   sys.<  (sys.SMALLDATETIME, date),
+    OPERATOR    2   sys.<= (sys.SMALLDATETIME, date),
+    OPERATOR    3   sys.=  (sys.SMALLDATETIME, date),
+    OPERATOR    4   sys.>= (sys.SMALLDATETIME, date),
+    OPERATOR    5   sys.>  (sys.SMALLDATETIME, date),
+    FUNCTION    1   sys.smalldatetime_date_cmp(sys.SMALLDATETIME, date);
+END IF;
+END $$;
+
+-- Opartor class for smalldatetime_ops to incorporate various operator between date and smalldatetime for Index scan
```
Comment (rohit01010) [inline contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql]: NIT: ~Opartor~ Operator

**Review by rohit01010 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql (hunk 10 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql]: resolved

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql (hunk 9 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql]: resolved

File: contrib/babelfishpg_common/sql/smalldatetime.sql (hunk 8 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/smalldatetime.sql:504]: Create a operators when left arg is date and right arg is sys.SMALLDATETIME in smalldatetime_ops family.

**Review by rohit01010 — APPROVED** [review]

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+\echo Use "ALTER EXTENSION ""babelfishpg_common"" UPDATE TO "4.5.0"" to load this file. \quit
+
+SELECT set_config('search_path', 'sys, '||current_setting('search_path'), false);
+
+CREATE OR REPLACE FUNCTION  sys.smalldatetime_date_cmp(sys.SMALLDATETIME, date)
+RETURNS INT4
+AS 'timestamp_cmp_date'
+LANGUAGE internal IMMUTABLE STRICT PARALLEL SAFE;
+
+CREATE FUNCTION  sys.date_smalldatetime_cmp(date, sys.SMALLDATETIME)
```
Comment (Deepesh125) [inline contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql]: we should be able to run any script repeatedly. So use `create or replace`

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out [hunk 12]
```diff
@@ -0,0 +1,3863 @@
...
+~~END~~
+
+
+SELECT test_smalldatetime_date_f1();
+GO
+~~START~~
+text
+Query Text: SELECT test_smalldatetime_date_f1()
+Result
+~~END~~
```
Comment (Deepesh125) [inline test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out]: no need for such testing

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out [hunk 13]
```diff
@@ -0,0 +1,3863 @@
...
+
+
+SELECT COUNT(*) FROM date_smalldatetime_cmp_t1 WHERE date_col BETWEEN CAST('2023-06-16' AS SMALLDATETIME) AND CAST('2023-06-18' AS SMALLDATETIME);
+GO
+~~START~~
+text
+Query Text: SELECT COUNT(*) FROM date_smalldatetime_cmp_t1 WHERE date_col BETWEEN CAST('2023-06-16' AS SMALLDATETIME) AND CAST('2023-06-18' AS SMALLDATETIME)
+Aggregate
+  ->  Seq Scan on date_smalldatetime_cmp_t1
+        Filter: ((date_col >= '2023-06-16 00:00:00'::smalldatetime(0) without time zone) AND (date_col <= '2023-06-18 00:00:00'::smalldatetime(0) without time zone))
```
Comment (Deepesh125) [inline test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out:1991]: can you also construct predicate which will choose index scan?

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out [hunk 14]
```diff
@@ -0,0 +1,3863 @@
...
+
+
+SELECT COUNT(*) FROM date_smalldatetime_cmp_t1 WHERE CAST('2023-06-14' AS SMALLDATETIME) >= date_col;
+GO
+~~START~~
+text
+Query Text: SELECT COUNT(*) FROM date_smalldatetime_cmp_t1 WHERE CAST('2023-06-14' AS SMALLDATETIME) >= date_col
+Aggregate
+  ->  Seq Scan on date_smalldatetime_cmp_t1
+        Filter: ('2023-06-14 00:00:00'::smalldatetime(0) without time zone >= date_col)
```
Comment (Deepesh125) [inline test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out:2023]: we should have tests where index scan is being picked

File: contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.8.0--4.0.0.sql [hunk 15]
```diff
@@ -0,0 +1,2 @@
+-- complain if script is sourced in psql, rather than via ALTER EXTENSION
```
Comment (Deepesh125) [inline contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.8.0--4.0.0.sql:1]: why is this needed now?

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out [hunk 16]
```diff
@@ -0,0 +1,3863 @@
...
+~~END~~
+
+~~START~~
+text
+Babelfish T-SQL Batch Parsing Time: 0.174 ms
+~~END~~
+
+
+-- index scan
+SELECT COUNT(*) FROM smalldate_date_cmp_t1 WHERE smalldatetime_col < CAST('2023-06-14' AS DATE);
```
Comment (Deepesh125) [inline test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out:129]: can you also have order by clause and confirm it still picks index scan?

Comment (Deepesh125) [reply test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out:129]: is it possible to add test which involes join of two tables where inner side relation picks up index scan

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out (hunk 13 above)
Comment (Deepesh125) [reply test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out:1991]: ignore this ^. can you please point out similar cases for smalldatetime_date_cmp_t1 tables?

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out (hunk 14 above)
Comment (Yvinayak07) [reply test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out:2023]: This is the the testing where we create index on date and then compare smalldatetime and date. 
Already filed [jira](https://jira.rds.a2z.com/browse/BABEL-5428) for this.

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out (hunk 13 above)
Comment (Yvinayak07) [reply test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out:1991]: Similar [case](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3085/files#diff-47f3942eb8058b6cdd3ddbbb1bc645d7726a6993d7ccff1768d3ddd43c0264bdR225).

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out (hunk 12 above)
Comment (Yvinayak07) [reply test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out]: Removed these tests.

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql (hunk 11 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.4.0--4.5.0.sql]: added create or replace.
Resolved

File: test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out (hunk 16 above)
Comment (Yvinayak07) [reply test/JDBC/expected/smalldatetime_date_cmp-vu-verify.out:129]: added tests
- ORDER BY clause.
- joins of 2 tables choosing index scan.

File: contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.8.0--4.0.0.sql (hunk 15 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.8.0--4.0.0.sql:1]: As we need to put this to 3_X and in 3_X we will have 3.7 -> 3.8  then in 4_X we should have 3.8->4.0.

**Review by sumitj824 — APPROVED** [review]

**Review by Deepesh125 — APPROVED** [review]

**Review by jsudrik — APPROVED** [review]
Approved.
